text/0000-colocated-tests.md

```diff
@@ -0,0 +1,161 @@
+ - now
+   - include: `app/**/*.{js,hbs}`
+   - exclude: `app/**/*.test.{js,ts}`
+
+For the test bundle,
+ - previously
+   - include: `tests/**/*.{js}`
+ - now
+   - include: `tests/**/*.{js}`
+   - include: `app/**/*.test.{js}`
+
+
+Like with component co-location, there will be different preferences with how to group files together. It may be undesirable to have a flat list of 20 routes, and then an additional flat list of 20 tests for those routes all in the same folder.
+The `.ember-cli` file provides the option to further nest components' class and template files as `app/components/{component-name}/index.{js,hbs}` via the `componentStructure` property (set to `nested`).
+This same flexibility should be available for routes, controllers, etc -- but is out of scope for this RFC as other resolver implementations may be in-flight.
+Avoiding implementing `componentStructure` for the rest of the `app` directory does not have any impact on test co-location as the resolver is not involved with finding test file locations. For example, it would be valid to have `app/routes/{route-name}.js` and tests located at `app/routes/__tests__/{route-name}.test.js`.
+
+
+### Reference: Where do the existing tests get migrated to?
+While tests in `app/` could live anywhere, these are to be the default locations for the generators to use.
+
+#### Route
+
+previously:
+ - `app/routes/{route-name}.js`
+ - `tests/unit/{route-name}-test.js`
+now:
+ - `app/routes/{route-name}.js`
+ - `app/routes/{route-name}.test.js`
+
+#### Service
+
+previously:
+ - `app/services/{service-name}.js`
+ - `tests/unit/{service-name}-test.js`
+now:
+ - `app/service/{service-name}.js`
+ - `app/service/{service-name}.test.js`
+
+#### Controller
+
+previously:
+ - `app/controllers/{controller-name}.js`
+ - `tests/unit/{controller-name}-test.js`
+now:
+ - `app/controllers/{controller-name}.js`
+ - `app/controllers/{controller-name}.test.js`
+
+#### Component
+
+previously:
+ - `app/components/{component-name}.js`
+ - `app/components/{component-name}.hbs`
+ - `tests/integration/components/{component-name}-test.js`
+now:
+ - `app/components/{component-name}.js`
+ - `app/components/{component-name}.hbs`
+ - `app/components/{component-name}.test.js`
+
+or using `"componentStructure": "nested"`:
+ - `app/components/{component-name}/index.js`
+ - `app/components/{component-name}/index.hbs`
+ - `app/components/{component-name}/index.test.js`
+
+#### Helpers
+
+previously:
+ - `app/helpers/{helper-name}.js`
+ - `tests/integration/helpers/{helper-name}-test.js`
+now:
+ - `app/helpers/{helper-name}.js`
+ - `app/helpers/{helper-name}.test.js`
+
+#### Utils
+
+previously:
+ - `app/utils/{util-name}.js`
+ - `tests/unit/{util-name}-test.js`
+now:
+ - `app/utils/{util-name}.js`
+ - `app/utils/{util-name}.test.js`
+
+#### Acceptance Tests
+
+Acceptance tests do not have a core affiliation with any other file, so co-locating them _may_ not make sense, depending on what a particular test file is doing.
```

Comment:
in the interest of not leaving a class of tests alone in the tests folder which would really act to minimize their importance, I would propose a default location for acceptance tests under the `app/routes/` folder

reasoning behind that is that in those tests you start off by visiting a route then testing the behavior of it as a whole

Reply:
is this how most acceptance tests are? I thought to leave them in `tests/acceptance/` because I haven't seen a clear pattern (not that I've seen a ton of different people's / companies' acceptance tests)

Reply:
that is partly why I left this comment
also judging by the guide https://guides.emberjs.com/release/tutorial/part-1/automated-testing/
that's what an acceptance test looks like:
`await visit('/')` click around assert stuff

Reply:
The fact that acceptance tests won't be co-located makes this proposal significantly weaker. Now there are two places to look for tests.

Reply:
I started a poll on twitter: https://twitter.com/tommyjr/status/1234563047873708032

my thoughts are -- what would they be co-located with?
you can co-locate them yourself, as all .test.js files in the app folder would be detected, but I have a hard time justifying changing the default for acceptance tests.

Reply:
collocating acceptance tests with routes would not work. We have several application level tests where the entry URL doesn't matter, we're just using it as a shell to test some complex interaction / behavior.

Reply:
I'm personally on the side of co-locating acceptance test with wherever the route lives. Most of the acceptance tests I've written do have some entry URL. I think it's the "sane default" here. And IMO this does not break other people's workflows. If it does not work for your case, you can still use the old locations.

Will the generator be able to use old locations? With some flag?

Also note that since the crawling code will use `**/*.test.js` we can use any location we like.

Reply:
The current acceptance test blueprint is pretty tied to a single route. I suppose you could then navigate to other routes, but then the auto-generated module and filename would not be very clear, etc. Whenever I create an acceptance test that isn't just/primarily testing a single route, I don't use the blueprint.

So I could definitely see adding another blueprint, e.g. `ember g route-acceptance-test my-route` vs. `ember g acceptance-test test-my-whole-app`, but there is clearly a need for a single-route-focused blueprint, and I do think it makes sense to co-locate that one.

So I think my opinion is:
1. Co-locate acceptance tests as the blueprint currently understands them
2. Maybe add another blueprint for non-route-focused blueprints that probably get generated in `tests/<somewhere>`.